src/rtc.rs

```diff
@@ -0,0 +1,172 @@
+use crate::rcc::Rcc;
+use crate::stm32::RTC;
+use crate::time::{Date, Time, U32Ext};
+
+/**
+ Interface to the real time clock
+**/
+
+pub struct Rtc {
+    regs: RTC,
+}
+
+impl Rtc {
+    /**
+    Initializes the RTC
+     **/
+    // Sets default clock source to LSI, since LSE is not included on STM32f3 discovery boards
+    pub fn rtc(regs: RTC, src: RTCSrc, rcc: &mut Rcc) -> Self {
+        let mut result = Rtc { regs };
+
+        rcc.enable_rtc(&src);
+
+        result.regs.cr.modify(|_, w| {
+            w
+                // sets hour format to 24 hours
+                .fmt()
+                .clear_bit()
+        });
+
+        // Prescalers set to produce a 1 hz signal
+        let (prediv_s, prediv_a) = match src {
```

Comment:
These numbers came from AN4795:

![temp](https://user-images.githubusercontent.com/5498472/80778524-8f237700-8b1d-11ea-94b8-5ca917932c53.png)

I like the idea of a function that takes care of choosing the correct prescalers to produce a 1 hz RTC clock source. This is possible for LSI and LSE, but the fact that HSE frequency can vary (I read that the HSE frequency can be between 4 and 32 Mhz) makes that a challenge. 

The concept of making some methods "visible" given a parameter (the RTCSrc) sounds like a good solution, if it is possible. 

For example :

if RTCSrc::LSE or RTCSrc::LSI, make the following method visible....

```
fn set_1hz_frequency{
   /// set prescalers to produce 1hz clock source
}
```

else, there is a more general function `set_frequency` that is always visible that would be used if the user is using an HSE of any frequency. The user would then supply prescalers to set appropriate frequency. 

 I think I am not using the correct words to search to see if this feature of making methods "visible" or "invisible" exists, because I am not able to find anything online. Can something like this be accomplished?

Reply:
Ah interesting to know. I would like to see a comment or make these tuples for fixed "input" frequencies (meaning LSI LSE) constants. 

I have difficulties to find the AN4795 Application note, can you provide a link?

Reply:
It was a typo. I meant AN4759
Here is the link:
[https://www.st.com/resource/en/application_note/dm00226326-using-the-hardware-realtime-clock-rtc-and-the-tamper-management-unit-tamp-with-stm32-microcontrollers-stmicroelectronics.pdf](url)

Reply:
Because HSE clock frequencies can vary, perhaps the user should choose a source and prescaler values to produce the desired rtc input clock speed through a configuration structure. 

eg.

```
// rtc module


/// RTC clock input source
#[derive(Clone, Copy)]
pub enum RTCSrc {
    LSE = 0b01,
    LSI = 0b10,
    HSE = 0b11,
}

/// RTC configuration struct
pub struct cfgRTC {
    src: RTCSrc,
    prediv_s: u16,
    prediv_a: u8,
}

impl cfgRTC {
    pub fn new(source: RTCSrc, prescaler_sync: u16, prescaler_async: u8) 
        -> cfgRTC 
        {
        let config  = cfgRTC {
            src: source,
            prediv_s: prescaler_sync,
            prediv_a: prescaler_async,
        };
        config
    }
}

pub struct Rtc {
    regs: RTC,
}

impl Rtc {
    pub fn rtc(regs: RTC, cfg: cfgRTC, rcc: &mut Rcc) -> Self {
        /// initialization of rtc with desired clock source and input frequency
    }
// etc etc
}

```